src/main/java/net/fabricmc/loom/configuration/CompileConfiguration.java

```diff
@@ -235,20 +240,91 @@ private Path getLockFile() {
 		return cacheDirectory.resolve("." + pathHash + ".lock");
 	}
 
-	private boolean getAndLock() {
-		final Path lock = getLockFile();
+	enum LockResult {
+		// acquired immediately or after waiting for another process to release
+		ACQUIRED_CLEAN,
+		// already owned by current pid
+		ACQUIRED_ALREADY_OWNED,
+		// acquired due to current owner not existing
+		ACQUIRED_PREVIOUS_OWNER_MISSING
+	}
 
-		if (Files.exists(lock)) {
-			return true;
-		}
+	private LockResult acquireProcessLockWaiting(Path lockFile) {
+		// one hour
+		return this.acquireProcessLockWaiting(lockFile, 1000L * 60 * 60);
+	}
 
+	private LockResult acquireProcessLockWaiting(Path lockFile, long timeoutMs) {
 		try {
-			Files.createFile(lock);
-		} catch (IOException e) {
-			throw new UncheckedIOException("Failed to acquire getProject() configuration lock", e);
+			return this.acquireProcessLockWaiting_(lockFile, timeoutMs);
+		} catch (final IOException e) {
+			throw new RuntimeException("Exception acquiring lock " + lockFile, e);
+		}
+	}
+
+	// Returns true if our process already owns the lock
+	@SuppressWarnings("BusyWait")
+	private LockResult acquireProcessLockWaiting_(Path lockFile, long timeoutMs) throws IOException {
+		final Logger logger = Logging.getLogger("loom_acquireProcessLockWaiting");
+		final long currentPid = ProcessHandle.current().pid();
+		boolean abrupt = false;
+
+		if (Files.exists(lockFile)) {
+			long lockingProcessId;
+
+			try {
+				lockingProcessId = Long.parseLong(Files.readString(lockFile));
+			} catch (final Exception e) {
+				lockingProcessId = -1;
+			}
+
+			if (lockingProcessId == currentPid) {
+				return LockResult.ACQUIRED_ALREADY_OWNED;
+			}
+
+			logger.lifecycle("Lock file '{}' is currently held by pid '{}'.", lockFile, lockingProcessId);
+
+			if (ProcessHandle.of(lockingProcessId).isEmpty()) {
+				logger.lifecycle("Locking process does not exist, assuming abrupt termination and deleting lock file.");
+				Files.deleteIfExists(lockFile);
+				abrupt = true;
+			} else {
+				logger.lifecycle("Waiting for lock to be released...");
+				long sleptMs = 0;
+
+				while (Files.exists(lockFile)) {
+					try {
+						Thread.sleep(100);
+					} catch (final InterruptedException e) {
+						Thread.currentThread().interrupt();
+					}
+
+					sleptMs += 100;
+
+					if (sleptMs >= 1000 * 60 && sleptMs % (1000 * 60) == 0L) {
+						logger.lifecycle(
+								"""
+										Have been waiting on lock file '{}' held by pid '{}' for {} minute(s).
+										If this persists for an unreasonable length of time, kill this process, run './gradlew --stop' and then try again.
+										If the problem persists, the lock file may need to be deleted manually.""",
```

Comment:
This looks great overall, I'm not super happy about suggesting to the user to delete the lock file. The issue with doing this is loom will contiune thinking everything is ok. In what situation may you need to delete the lockfile?

Reply:
Maybe the soltuion is to write -1 to the lockfile here, so any future runs will go down the route of the previous owner missing?

Reply:
I think simply changing the message is enough, if we were to write `-1` while a zombie process holds the lock it could cause other problems...

Reply:
Ok, thats fine. If it does become an issue we could possibly look into getting more info about the other process (such as its command line, or whatever to try and figure out what it is). Not sure how easy that is to do so dont worry about it for this PR.